File: cole/day13_diameter-of-binary-tree/diameter-of-binary-tree.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::cell::RefCell;
use std::cmp::max;
use std::rc::Rc;
fn d(root: Option<Rc<RefCell<TreeNode>>>) -> (i32, i32) {
    match root {
        Some(r) => {
            let (ld, lP) = d(r.borrow().left.clone());
            let (rd, rP) = d(r.borrow().right.clone());
            (max(ld, max(rd, lP + rP)), max(lP, rP) + 1)
        }
        None => (0, 0),
    }
}
impl Solution {
    pub fn diameter_of_binary_tree(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        let (d, _) = d(root);
        d
    }
}

File: gabe/diameter_of_binary_tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        return max(self.treeDiameter(root))

        
    def treeDiameter(self, root):
        if root == None:
            return (-1, 0)

        left_max, whole_left_max = self.treeDiameter(root.left)
        right_max, whole_right_max = self.treeDiameter(root.right)

        side_max = max(left_max, right_max) + 1
        whole_max = max(whole_left_max, whole_right_max, left_max + right_max + 2)

        return (side_max, whole_max)